spec/support/authn_k8s/authn_k8s_test_server.rb

```diff
@@ -0,0 +1,152 @@
+
+            # Below is an example of the server writing to the client via websockets
+            # ws.send("well then :)".bytes.unshift(1))
+            ws.close(1000)
+        end
+        
+        ws.on :message do |event|
+            type, *message_bytes = event.data.bytes
+            message = message_bytes.pack('c*')
+            log_event([:message, type])
+            # log_event([:message, type, message])
+
+            path, content, mode = AuthnK8sTestServer.match_script(message)
+            @copied_content = content
+            # AuthnK8sTestServer.log("Updated @copied_content: \n#{copied_content}")
+            ws.close(1000)
+        end
+    
+    
+        ws.on :close do |event|
+            log_event([:close, event.code, event.reason])
+          ws = nil
+        end
+    
+        # Return async Rack response
+        ws.rack_response
+    end
+
+    def call(env)
+        req = Rack::Request.new(env)
+        AuthnK8sTestServer.log("Handling request: #{req.request_method} path=#{req.fullpath}")
+
+        if env['HTTP_AUTHORIZATION'] != "Bearer #{bearer_token}"
+            return [ 401, {"Content-Type" => "application/json"}, [AuthnK8sTestServer.read_response_file("unauthorized.json")] ]
+        end
+
+        return ws_call(env) if Faye::WebSocket.websocket?(env)
+
+        if req.path == "#{subpath}/api/v1"
+            [ 200, {"Content-Type" => "application/json"}, [AuthnK8sTestServer.read_response_file("good:api.v1.json")] ]
+        elsif req.path == "#{subpath}/api"
+            [ 200, {"Content-Type" => "application/json"}, [AuthnK8sTestServer.read_response_file("good:api.json")] ]
+        elsif req.path == "#{subpath}/apis"
+            [ 200, {"Content-Type" => "application/json"}, [AuthnK8sTestServer.read_response_file("good:apis.json")] ]
+        elsif req.path.start_with?("#{subpath}/apis/")
+            [ 200, {"Content-Type" => "application/json"}, [AuthnK8sTestServer.read_response_file("good:apis.all.json")] ]
+        elsif req.path == "#{subpath}/api/v1/namespaces/default/pods/bash-8449b79d7-c2fwd"
+            [ 200, {"Content-Type" => "application/json"}, [AuthnK8sTestServer.read_response_file("good:api.v1.getpod.json")] ]
+        elsif req.path.start_with?("#{subpath}/api/v1/namespaces/default/pods/")
+            [ 404, {"Content-Type" => "application/json"}, [AuthnK8sTestServer.read_response_file("bad:api.v1.getpod.json")] ]
+        elsif req.fullpath == "#{subpath}/api/v1/namespaces?labelSelector=field.cattle.io%2FprojectId%3Dp-q7s7z&fieldSelector=metadata.name%3Ddefault"
+            [ 200, {"Content-Type" => "application/json"}, [AuthnK8sTestServer.read_response_file("good:api.v1.getnamespaces.json")] ]
+        elsif req.path.start_with?("#{subpath}/api/v1/namespaces")
+            [ 200, {"Content-Type" => "application/json"}, [AuthnK8sTestServer.read_response_file("bad:api.v1.getnamespaces.json")] ]
+        else
+            [ 200, {'Content-Type' => "application/json"}, ['{}'] ]
```

Comment:
Might be a nitpick, but as we're enumerating all the endpoints the test case should need, maybe our fallback `else` condition should return a `40x` error instead of `200` response.

Reply:
This is a fair point. The issue I faced is that clients for Kubernetes hit a whole list of endpoints for (I think) service discovery purposes at initialisation time. Having the endpoints in the else branch return 200 overcame this issue. There's probably a better way.

I'll leave a comment on the else branch and then we can make improvements during the upcoming iterations on the test server.